File: discrete_model.py
#------------------------------------------------------------------------------
# discrete_model python module
# (using partition method for solving linear equations)
# 
# ricardoaf@lccv.ufal.br (version 17/11/2020)
#------------------------------------------------------------------------------
import numpy as np

#------------------------------------------------------------------------------
def run (node, elem, alpha):

	# global stiffness matrix
	kg = kglobal (elem)

	# partition method
	node = partition (node, kg)

	# eval element forces
	elem = internalf (node, elem, alpha)

	# return nodes and elements
	return node, elem

#------------------------------------------------------------------------------
def kglobal (elem):

	nelem = len(elem['k'])
	nnode = max(elem['c'].flatten())

	# init matrix
	k = np.zeros((nnode, nnode))

	# element loop
	for i in range(nelem):

		# local matrix
		ke = elem['k'][i] * np.array([[1,-1], [-1, 1]])

		# global assemble
		conn = elem['c'][i]-1
		k[np.ix_(conn, conn)] += ke

	# return matrix
	return k

#------------------------------------------------------------------------------
def partition (node, k):

	# free nodes
	fix = node['fix']-1
	dof = np.setdiff1d(np.arange(len(node['u'])), fix)

	# sub matrices
	kdd, kdf, kff = k[np.ix_(dof,dof)], k[np.ix_(dof,fix)], k[np.ix_(fix,fix)]

	# solve unknowns
	node['u'][dof] = np.linalg.solve(kdd, node['f'][dof] - kdf @ node['u'][fix])
	node['f'][fix] = kdf.T @ node['u'][dof] + kff @ node['u'][fix]

	# return nodes
	return node

#------------------------------------------------------------------------------
def internalf (node, elem, alpha):

	# i and j nodes from all elements
	i = elem['c'][:, 0]-1
	j = elem['c'][:, 1]-1

	# calc relative u and element f
	du = node['u'][np.ix_(j)] - node['u'][np.ix_(i)]
	elem['f'] = alpha * elem['k'] * du

	# return elements
	return elem

File: julia/DiscreteModel.jl
#------------------------------------------------------------------------------
# discrete_model julia file
# (using partition method for solving linear equations)
# 
# ricardoaf@lccv.ufal.br (version 21/01/2021)
#------------------------------------------------------------------------------

struct Node; u::Vector{Float64}; f::Vector{Float64}; fix::Vector{Int}; end
struct Elem; c::Array{Int,2}; k::Vector{Float64}; f::Vector{Float64}; end

function init(;connectivity, stiffness, displacement, force, fixnode, type="mechanic")
	Node(Float64.(displacement), Float64.(force), Int.(fixnode)),
	Elem(Int.(connectivity), Float64.(stiffness), zeros(length(stiffness))),
	type ≢ "hydraulic" ? 1. : -1.
end

function run!(node, elem, α)
	# global stiffness matrix and partition method	
	partition!(node, kglobal(elem))
	# eval element forces
	internalf!(elem, node, α)
	# return
	node, elem
end

function kglobal(elem)
	# number of nodes and elements
	nnode, nelem = maximum(elem.c), length(elem.k)
	# init matrix
	k = zeros(nnode, nnode)
	# element loop: global assemble from local matrix
	for i = 1:nelem
		k[elem.c[i, :], elem.c[i, :]] += elem.k[i] * [1 -1; -1 1.]
	end
	k # return k
end

function partition!(node, k)
	# free nodes
	fix, dof = node.fix, setdiff(1:length(node.u), node.fix)
	# solve unknowns
	node.u[dof] = k[dof, dof] \ (node.f[dof] - k[dof, fix] * node.u[fix])
	node.f[fix] = k[fix, dof] * node.u[dof] + k[fix, fix] * node.u[fix]
end

function internalf!(elem, node, α)
	# i and j nodes from all elements
	i, j = elem.c[:, 1], elem.c[:, 2]
	# calc element force from relative disp
	@. elem.f = α * elem.k * (node.u[j] - node.u[i])
end
